libs/utils/analysis/idle_analysis.py

```diff
@@ -0,0 +1,266 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2015, ARM Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+""" Idle Analysis Module """
+
+import matplotlib.gridspec as gridspec
+import matplotlib.pyplot as plt
+import pandas as pd
+import pylab as pl
+
+from analysis_module import AnalysisModule
+from trace import ResidencyTime, ResidencyData
+from trappy.utils import listify
+
+# Configure logging
+import logging
+
+
+class IdleAnalysis(AnalysisModule):
+    """
+    Support for plotting Idle Analysis data
+
+    :param trace: input Trace object
+    :type trace: :mod:`libs.utils.Trace`
+    """
+
+    def __init__(self, trace):
+        super(IdleAnalysis, self).__init__(trace)
+
+###############################################################################
+# DataFrame Getter Methods
+###############################################################################
+
+    def _dfg_idle_state_residency(self, entity):
+        """
+        Compute time spent by a given entity in each idle state.
+
+        :param entity: cpu ID or cluster name or list of CPU IDs
+        :type entity: int or str
+
+        :returns: :mod:`pandas.DataFrame` - idle state residency dataframe
+        """
+        if not self._trace.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'idle state residency computation not possible!')
+            return None
+
+        if isinstance(entity, str):
+            try:
+                _entity = self._platform['clusters'][entity.lower()]
+            except KeyError:
+                logging.warn('%s entity not found!', entity)
+                return None
+        else:
+            _entity = listify(entity)
+
+        idle_df = self._dfg_trace_event('cpu_idle')
+        entity_idle_df = idle_df[idle_df.cpu_id == _entity[0]]
```

Comment:
I think that's a mistake that came up while trying to mimic the residency computation from frequency analysis... In that case, since we assume frequency coherency in the cluster, it is enough to pick data related to a single CPU in the cluster. For idle, this is not true, CPUs can be in different idle states. I'll change this accordingly...